autocodesign/certificates.go

```diff
@@ -0,0 +1,219 @@
+
+	for certificateType, required := range requiredCertificateTypes {
+		if required && len(typeToLocalCerts[certificateType]) == 0 {
+			return map[appstoreconnect.CertificateType][]Certificate{}, missingCertificateError{certificateType, teamID}
+		}
+	}
+
+	// only for debugging
+	if isDebugLog {
+		if err := logAllAPICertificates(client, typeToLocalCerts); err != nil {
+			log.Debugf("Failed to log all Developer Portal certificates: %s", err)
+		}
+	}
+
+	validAPICertificates := map[appstoreconnect.CertificateType][]Certificate{}
+	for certificateType, validLocalCertificates := range typeToLocalCerts {
+		matchingCertificates, err := matchLocalToAPICertificates(client, certificateType, validLocalCertificates)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(matchingCertificates) > 0 {
+			log.Debugf("Certificates type %s has matches on Developer Portal:", certificateType)
+			for _, cert := range matchingCertificates {
+				log.Debugf("- %s", cert.Certificate)
+			}
+		}
+
+		if requiredCertificateTypes[certificateType] && len(matchingCertificates) == 0 {
+			return nil, fmt.Errorf("not found any of the following %s certificates on Developer Portal:\n%s", certificateType, certsToString(validLocalCertificates))
+		}
+
+		if len(matchingCertificates) > 0 {
+			validAPICertificates[certificateType] = matchingCertificates
+		}
+	}
+
+	return validAPICertificates, nil
+}
+
+// getValidLocalCertificates returns validated and deduplicated local certificates
+func getValidLocalCertificates(certificates []certificateutil.CertificateInfoModel, teamID string) (map[appstoreconnect.CertificateType][]certificateutil.CertificateInfoModel, error) {
+	preFilteredCerts := certificateutil.FilterValidCertificateInfos(certificates)
+
+	if len(preFilteredCerts.InvalidCertificates) != 0 {
+		log.Warnf("Ignoring expired or not yet valid certificates: %s", preFilteredCerts.InvalidCertificates)
+	}
+	if len(preFilteredCerts.DuplicatedCertificates) != 0 {
+		log.Warnf("Ignoring duplicated certificates with the same name: %s", preFilteredCerts.DuplicatedCertificates)
+	}
+
+	log.Debugf("Valid and deduplicated certificates:\n%s", certsToString(preFilteredCerts.ValidCertificates))
+
+	localCertificates := map[appstoreconnect.CertificateType][]certificateutil.CertificateInfoModel{}
+	for _, certType := range []appstoreconnect.CertificateType{appstoreconnect.IOSDevelopment, appstoreconnect.IOSDistribution} {
+		localCertificates[certType] = filterCertificates(preFilteredCerts.ValidCertificates, certType, teamID)
+	}
+
+	log.Debugf("Valid and deduplicated certificates for Development team (%s):\n%s", teamID, certsToString(preFilteredCerts.ValidCertificates))
+
+	return localCertificates, nil
+}
+
+// matchLocalToAPICertificates ...
+func matchLocalToAPICertificates(client DevPortalClient, certificateType appstoreconnect.CertificateType, localCertificates []certificateutil.CertificateInfoModel) ([]Certificate, error) {
```

Comment:
`certificateType` is unused, are you sure it's working as intended?

Reply:
Removed unused param